sql/core/src/main/scala/org/apache/spark/sql/execution/ui/SQLAppStatusListener.scala

```diff
@@ -0,0 +1,353 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.sql.execution.ui
+
+import java.util.Date
+import java.util.concurrent.ConcurrentHashMap
+
+import scala.collection.JavaConverters._
+import scala.collection.mutable.HashMap
+
+import org.apache.spark.{JobExecutionStatus, SparkConf}
+import org.apache.spark.internal.Logging
+import org.apache.spark.scheduler._
+import org.apache.spark.sql.execution.SQLExecution
+import org.apache.spark.sql.execution.metric._
+import org.apache.spark.sql.internal.StaticSQLConf._
+import org.apache.spark.status.LiveEntity
+import org.apache.spark.status.config._
+import org.apache.spark.ui.SparkUI
+import org.apache.spark.util.kvstore.KVStore
+
+private[sql] class SQLAppStatusListener(
```

Comment:
I don't see calls to `update` in `onTaskEnd` or `onExecutorMetricsUpdate`.  Does that mean the live UI wont' update till a stage is finished?  But after looking at the tests, I guess I'm wrong, it does update ... where is the update I'm missing?

Reply:
The live UI is updated from the info in `LiveStageMetrics`, which is not written to the store. That's kept in memory while executions are running, and aggregated into the final metrics view when the execution finishes (see `aggregateMetrics`).

`SQLAppStatusStore.executionMetrics` has logic to call the listener directly when the final metrics are not yet computed.

Reply:
> SQLAppStatusStore.executionMetrics has logic to call the listener directly when the final metrics are not yet computed.

ahh, that is the part I hadn't noticed.  thanks